src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,423 @@
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.callbackRepository = callbackRepository;
+            this.ruleRepository = ruleRepository;
+            this.callbackExecuter = callbackExecuter;
+            this.blocks = blocks;
+            this.logger = logger;
+
+            this.watcher = new Zcoin.Watching.TransactionConfirmationWatcher<Rule>
+            (
+                this,
+                blocks
+            );
+
+            this.timerLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+            this.timers = new Dictionary<uint256, Dictionary<Guid, Timer>>();
+            this.watches = new ConcurrentDictionary<Guid, TransactionWatch<Rule>>();
+        }
+
+        public async Task<Rule> AddTransactionAsync(
+            uint256 transaction,
+            int confirmation,
+            TimeSpan unconfirmedWaitingTime,
+            Callback callback,
+            TransactionConfirmationCallbackResult successData,
+            TransactionConfirmationCallbackResult timeoutData,
+            CancellationToken cancellationToken)
+        {
+            if (transaction == null)
+            {
+                throw new ArgumentNullException(nameof(transaction));
+            }
+
+            if (confirmation < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(confirmation) ,"Confirmation is less than zero.");
+            }
+
+            if (!Timer.IsValidDuration(unconfirmedWaitingTime))
+            {
+                throw new ArgumentOutOfRangeException(nameof(unconfirmedWaitingTime), "UnconfirmedWaitingTime is invalid duration.");
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            if (successData == null)
+            {
+                throw new ArgumentNullException(nameof(successData));
+            }
+
+            if (timeoutData == null)
+            {
+                throw new ArgumentNullException(nameof(timeoutData));
+            }
+
+            var rule = await this.ruleRepository.AddAsync
+            (
+                transaction,
+                confirmation,
+                unconfirmedWaitingTime,
+                successData,
+                timeoutData,
+                callback,
+                cancellationToken
+            );
+
+            await SetupTimerAsync(rule);
+
+            return rule;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var rules = await this.ruleRepository.ListAsync(cancellationToken);
+
+            foreach (var rule in rules.Where(w => !w.Completed))
+            {
+                var (block, _) = await this.blocks.GetByTransactionAsync(rule.Transaction, cancellationToken);
```

Comment:
This is a bug caused by another bug!. Due to there is a possibility that one transaction will live inside more than one blocks (and we already have this bugs too!).